internal/mdns/mdns.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package mdns
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/whyrusleeping/mdns"
+)
+
+// Notifee is notified when a new peer is found.
+type Notifee interface {
+	HandlePeerFound(peer.AddrInfo)
+}
+
+// Service implements a mDNS service.
+type Service struct {
+	server     *mdns.Server
+	p2pHost    IDNetworker
+	serviceTag string
+	notifee    Notifee
+	logger     Logger
+	started    bool
+	stop       chan struct{}
+	done       chan struct{}
+}
+
+// NewService creates and returns a new mDNS service.
+func NewService(p2pHost IDNetworker, serviceTag string,
+	logger Logger, notifee Notifee) (service *Service) {
+	if serviceTag == "" {
+		serviceTag = "_ipfs-discovery._udp"
+	}
+
+	return &Service{
+		p2pHost:    p2pHost,
+		serviceTag: serviceTag,
+		notifee:    notifee,
+		logger:     logger,
+	}
+}
+
+// Start starts the mDNS service.
+func (s *Service) Start() (err error) {
+	ips, port := getMDNSIPsAndPort(s.p2pHost)
+
+	hostID := s.p2pHost.ID()
+
+	hostIDPretty := hostID.Pretty()
+	txt := []string{hostIDPretty}
+	mdnsService, err := mdns.NewMDNSService(hostIDPretty, s.serviceTag, "", "", int(port), ips, txt)
+	if err != nil {
+		return fmt.Errorf("creating mDNS service: %w", err)
+	}
+
+	server, err := mdns.NewServer(&mdns.Config{Zone: mdnsService})
+	if err != nil {
+		return fmt.Errorf("creating mDNS server: %w", err)
+	}
+	s.server = server
+
+	s.started = true
+	s.stop = make(chan struct{})
+	s.done = make(chan struct{})
+	ready := make(chan struct{})
+
+	go s.run(ready)
+	// It takes a few milliseconds to launch a goroutine
+	// so we wait for the run goroutine to be ready.
+	<-ready
+
+	return nil
+}
+
+// Stop stops the mDNS service and server.
+func (s *Service) Stop() (err error) {
+	if !s.started {
+		return nil
+	}
+
+	defer func() {
+		s.started = false
+	}()
+	close(s.stop)
+	<-s.done
+	return s.server.Shutdown()
+}
+
+func (s *Service) run(ready chan<- struct{}) {
+	defer close(s.done)
+
+	const pollPeriod = time.Minute
```

Comment:
this should be configurable on the service itself.  If there were tests we wouldn't want to wait a minute.

Reply:
Indeed :+1: 
Moved as a struct field. I didn't expose it as argument to the service creation, since it wasn't exposed in `mdns_legacy` and we would lose access to it once upgrading to the newer `mdns` (when they fix their stuff)